File: apps/shared/src/library/hash.ts
import type { LibraryInvalidReason } from './schemas';

const textEncoder = new TextEncoder();

/**
 * Domain separators keep the two hash shapes in disjoint namespaces. Without them a file whose
 * bytes happen to spell a one-entry manifest (`<name>\0<64 hex>\n`) would hash identically to
 * the directory that manifest describes, and a file-backed and a directory-backed instance of
 * the same resource would be grouped as identical content.
 */
const FILE_HASH_DOMAIN = 'mangostudio/library/file\0';
/**
 * Versioned so the manifest-injectivity fix below changes every stored directory hash instead of
 * silently colliding with the pre-fix encoding: a length-prefixed manifest and a
 * `\0`/`\n`-delimited one can never be mistaken for each other once they hash into disjoint
 * domains.
 *
 * Advertised on the runtime capability manifest. File-backed resources are
 * unaffected — only this directory domain moved.
 */
export const DIRECTORY_HASH_DOMAIN = 'mangostudio/library/dir/v2\0';

/**
 * Peers that omit `directoryHashDomain` on hello. Directory hashing moved to
 * v2 (`mangostudio/library/dir/v2`) before this field existed, so silence is
 * that already-shipped domain rather than the pre-length-prefix v1 encoding.
 *
 * Keep this pinned at 2 when the live domain becomes v3: omitted still means
 * "built after v2 hashing and before the advertisement", not "whatever this
 * binary computes today". Pre-v2 binaries sit behind that breaking hash change
 * and cannot be distinguished on hello.
 */
export const DEFAULT_DIRECTORY_HASH_DOMAIN_VERSION = 2;

/**
 * Largest version the capability manifest and the lifecycle view accept, and
 * the bound this module enforces when deriving one. Both are the same limit on
 * purpose: a domain the runtime can hash with but not advertise would fail
 * schema validation on hello, taking the whole handshake down over a hash
 * bump. Failing here instead turns that into a test failure in this repo.
 */
export const MAX_DIRECTORY_HASH_DOMAIN_VERSION = 255;

const DIRECTORY_HASH_DOMAIN_VERSION_PATTERN = /\/v(\d+)\0$/;

/**
 * The integer the runtime advertises for {@link DIRECTORY_HASH_DOMAIN}. Derived
 * from the domain string so a v3 bump cannot forget to update the capability.
 */
export function directoryHashDomainVersion(domain: string = DIRECTORY_HASH_DOMAIN): number {
  const match = DIRECTORY_HASH_DOMAIN_VERSION_PATTERN.exec(domain);
  const version = match ? Number(match[1]) : Number.NaN;
  if (!Number.isSafeInteger(version) || version < 1) {
    throw new Error(
      `Directory hash domain must end in /v<n>\\0 so the runtime can advertise it (got ${JSON.stringify(domain)}).`
    );
  }
  if (version > MAX_DIRECTORY_HASH_DOMAIN_VERSION) {
    throw new Error(
      `Directory hash domain version ${version} exceeds the ${MAX_DIRECTORY_HASH_DOMAIN_VERSION} the runtime can advertise; widen the capability schema in the same change.`
    );
  }
  return version;
}

/** Absent on the wire means v2 — the domain that shipped before this field. */
export function directoryHashDomainOf(advertised: number | undefined): number {
  return advertised ?? DEFAULT_DIRECTORY_HASH_DOMAIN_VERSION;
}

/**
 * Which platform's path semantics apply to the paths a {@link LibraryHashReader} returns.
 *
 * `\` is a legal POSIX filename character, so rewriting it to `/` unconditionally invents a
 * separator that is not there and lets an unrelated directory satisfy a containment check. This
 * module cannot sniff the answer — `apps/shared` stays framework-agnostic and has no
 * `process.platform` to read — so the adapter that owns a real filesystem supplies it.
 */
export type LibraryHashPathStyle = 'posix' | 'win32';

export interface LibraryHashReader {
  /**
   * Returns every leaf file under `rootPath` as a POSIX-style relative path.
   * Directory symlinks must be traversed into leaf entries by the adapter.
   */
  listFiles(rootPath: string): ReadonlyArray<string> | Promise<ReadonlyArray<string>>;
  /** Resolves symlinks and returns the canonical absolute path. */
  realPath(path: string): string | Promise<string>;
  readFile(path: string): Uint8Array | Promise<Uint8Array>;
  readonly pathStyle: LibraryHashPathStyle;
}

export interface LibraryContentHash {
  readonly contentHash: string;
  readonly sizeBytes: number;
}

export type LibraryDirectoryHash =
  | (LibraryContentHash & { readonly valid: true })
  | { readonly valid: false; readonly invalidReason: LibraryInvalidReason };

export async function hashLibraryFile(
  path: string,
  reader: Pick<LibraryHashReader, 'readFile'>
): Promise<LibraryContentHash> {
  const bytes = await reader.readFile(path);
  return {
    contentHash: await sha256(FILE_HASH_DOMAIN, bytes),
    sizeBytes: bytes.byteLength,
  };
}

export async function hashLibraryDirectory(
  rootPath: string,
  reader: LibraryHashReader
): Promise<LibraryDirectoryHash> {
  const { pathStyle } = reader;
  const canonicalRoot = normalizeHashPath(await reader.realPath(rootPath), pathStyle);
  const relativePaths = [...(await reader.listFiles(rootPath))].sort(comparePaths);
  const manifestLines: string[] = [];
  let sizeBytes = 0;

  for (const relativePath of relativePaths) {
    const violation = relativePathViolation(relativePath, pathStyle);
    if (violation) return invalid(violation);

    const canonicalFile = normalizeHashPath(
      await reader.realPath(joinPath(canonicalRoot, relativePath)),
      pathStyle
    );
    if (!isPathWithin(canonicalRoot, canonicalFile, pathStyle)) return invalid('path-escape');

    const bytes = await reader.readFile(canonicalFile);
    const fileHash = await sha256(FILE_HASH_DOMAIN, bytes);
    // Length-prefixed so no filename can spell a manifest line: `relativePathViolation` below is
    // the policy that rejects a `\n`/`\0` name outright, and this is the property that holds even
    // if a future caller reaches the hasher another way.
    manifestLines.push(`${relativePath.length}:${relativePath}${fileHash}`);
    sizeBytes += bytes.byteLength;
  }

  return {
    contentHash: await sha256(DIRECTORY_HASH_DOMAIN, textEncoder.encode(manifestLines.join(''))),
    sizeBytes,
    valid: true,
  };
}

async function sha256(domain: string, bytes: Uint8Array): Promise<string> {
  const prefix = textEncoder.encode(domain);
  // The copy also narrows `Uint8Array<ArrayBufferLike>` to a `BufferSource` digest accepts.
  const digestInput = new Uint8Array(prefix.byteLength + bytes.byteLength);
  digestInput.set(prefix);
  digestInput.set(bytes, prefix.byteLength);
  const digest = new Uint8Array(await crypto.subtle.digest('SHA-256', digestInput));
  return Array.from(digest, (byte) => byte.toString(16).padStart(2, '0')).join('');
}

/**
 * Canonical spelling of an already-resolved path, so a caller that wants to recognize the paths
 * this module hands to `readFile` keys them the same way. Backslashes are only ever a separator
 * under `'win32'` — under `'posix'` one is an ordinary filename character and survives untouched.
 */
export function normalizeHashPath(path: string, pathStyle: LibraryHashPathStyle): string {
  const normalized = pathStyle === 'win32' ? path.replaceAll('\\', '/') : path;
  if (normalized === '/') return normalized;
  return stripTrailingSlashes(normalized);
}

/**
 * Strips trailing `/` in a single linear pass.
 *
 * Replaces a `/\/+$/` cleanup that backtracks on library-supplied paths with
 * many repetitions of `/` (CodeQL js/polynomial-redos).
 */
function stripTrailingSlashes(path: string): string {
  let end = path.length;
  while (end > 0 && path[end - 1] === '/') end -= 1;
  return path.slice(0, end);
}

function joinPath(rootPath: string, relativePath: string): string {
  return `${rootPath}/${relativePath}`;
}

function comparePaths(left: string, right: string): number {
  if (left < right) return -1;
  if (left > right) return 1;
  return 0;
}

/**
 * Why a relative path cannot be hashed: `path-escape` for anything that could climb out of the
 * root (absolute, `.`, `..`, a Windows drive prefix), `unsafe-name` for a name that is safe to
 * resolve but unsafe to put in the manifest verbatim (`\n` or `\0`, both legal POSIX filename
 * bytes that could otherwise forge a manifest line).
 */
function relativePathViolation(
  path: string,
  pathStyle: LibraryHashPathStyle
): 'path-escape' | 'unsafe-name' | undefined {
  if (path.length === 0 || path.startsWith('/')) return 'path-escape';
  if (pathStyle === 'win32' && /^[A-Za-z]:[\\/]/.test(path)) return 'path-escape';
  if (path.includes('\n') || path.includes('\0')) return 'unsafe-name';

  const segments = pathStyle === 'win32' ? path.replaceAll('\\', '/').split('/') : path.split('/');
  const hasBadSegment = segments.some(
    (segment) => segment.length === 0 || segment === '.' || segment === '..'
  );
  return hasBadSegment ? 'path-escape' : undefined;
}

function isPathWithin(
  rootPath: string,
  candidatePath: string,
  pathStyle: LibraryHashPathStyle
): boolean {
  const caseInsensitive = pathStyle === 'win32';
  const root = caseInsensitive ? rootPath.toLowerCase() : rootPath;
  const candidate = caseInsensitive ? candidatePath.toLowerCase() : candidatePath;
  return candidate === root || candidate.startsWith(`${root}/`);
}

function invalid(invalidReason: LibraryInvalidReason): LibraryDirectoryHash {
  return { valid: false, invalidReason };
}

File: crates/mangostudio-runtime/src/library/hash.rs
//! Library content hashes, byte-compatible with
//! `apps/shared/src/library/hash.ts` and the whitespace digests in
//! `apps/shared/src/library/machine/instance-reader.ts`.
//!
//! Three hash shapes live here, each in its own domain so no two can
//! collide: a file's bytes (`mangostudio/library/file\0`), a directory's
//! length-prefixed manifest (`mangostudio/library/dir/v2\0`, whose version the
//! capability manifest advertises as `directoryHashDomain`), and the
//! whitespace-insensitive digest the hub uses to tell a formatting-only
//! divergence from a real one.

use sha2::{Digest, Sha256};

use super::collation::locale_compare;
use super::js::{cmp_utf16, strip_js_whitespace, text_decoder_decode, utf16_len};

const FILE_HASH_DOMAIN: &[u8] = b"mangostudio/library/file\0";
/// `DIRECTORY_HASH_DOMAIN`. Versioned: v2 is the length-prefixed manifest.
pub(crate) const DIRECTORY_HASH_DOMAIN: &[u8] = b"mangostudio/library/dir/v2\0";

/// The version the capability manifest advertises as `directoryHashDomain`,
/// derived from [`DIRECTORY_HASH_DOMAIN`] at compile time so a v3 bump cannot
/// forget the advertisement (`directoryHashDomainVersion` in
/// `apps/shared/src/library/hash.ts` makes the same derivation).
///
/// Usage: `manifest.directory_hash_domain = Some(DIRECTORY_HASH_DOMAIN_VERSION)`.
pub(crate) const DIRECTORY_HASH_DOMAIN_VERSION: u32 =
    directory_hash_domain_version(DIRECTORY_HASH_DOMAIN);

/// The largest version the capability schema accepts
/// (`MAX_DIRECTORY_HASH_DOMAIN_VERSION` in `apps/shared/src/library/hash.ts`).
const MAX_DIRECTORY_HASH_DOMAIN_VERSION: u32 = 255;

/// Reads `<n>` out of a domain shaped `.../v<n>\0`; any other shape, or a
/// version the capability schema cannot carry, fails the build.
const fn directory_hash_domain_version(domain: &[u8]) -> u32 {
    assert!(
        domain.len() >= 4 && domain[domain.len() - 1] == 0,
        "the directory hash domain must end in /v<n>\\0"
    );
    let mut start = domain.len() - 1;
    while start > 0 && domain[start - 1].is_ascii_digit() {
        start -= 1;
    }
    assert!(
        start >= 2 && start < domain.len() - 1,
        "the directory hash domain must end in /v<n>\\0"
    );
    assert!(
        domain[start - 1] == b'v' && domain[start - 2] == b'/',
        "the directory hash domain must end in /v<n>\\0"
    );
    let mut version: u32 = 0;
    let mut index = start;
    while index < domain.len() - 1 {
        assert!(
            version <= MAX_DIRECTORY_HASH_DOMAIN_VERSION,
            "the directory hash domain version exceeds what the capability schema can advertise"
        );
        version = version * 10 + (domain[index] - b'0') as u32;
        index += 1;
    }
    assert!(
        version >= 1 && version <= MAX_DIRECTORY_HASH_DOMAIN_VERSION,
        "the directory hash domain version must be 1..=255 so the capability schema can advertise it"
    );
    version
}
const WHITESPACE_HASH_DOMAIN: &[u8] = b"mangostudio/library/whitespace\0";

/// Why a directory cannot be hashed, as `LibraryInvalidReason` spells it.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum ManifestViolation {
    /// A relative path that could climb out of the root.
    PathEscape,
    /// A relative path carrying `\n` or `\0`, which could forge a manifest line.
    UnsafeName,
}

fn sha256_hex(domain: &[u8], bytes: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(domain);
    hasher.update(bytes);
    hex(&hasher.finalize())
}

fn hex(bytes: &[u8]) -> String {
    use std::fmt::Write;
    bytes
        .iter()
        .fold(String::with_capacity(64), |mut out, byte| {
            let _ = write!(out, "{byte:02x}");
            out
        })
}

/// `hashLibraryFile`'s digest: SHA-256 over the file domain and the bytes.
#[must_use]
pub(crate) fn hash_file_bytes(bytes: &[u8]) -> String {
    sha256_hex(FILE_HASH_DOMAIN, bytes)
}

/// `relativePathViolation` for POSIX-style relative paths. The Windows drive
/// check applies only when `win32` is set, as `pathStyle` decides it in
/// TypeScript; backslashes are separators only then too.
#[must_use]
pub(crate) fn relative_path_violation(path: &str, win32: bool) -> Option<ManifestViolation> {
    if path.is_empty() || path.starts_with('/') {
        return Some(ManifestViolation::PathEscape);
    }
    let bytes = path.as_bytes();
    if win32
        && bytes.len() >= 3
        && bytes[0].is_ascii_alphabetic()
        && bytes[1] == b':'
        && (bytes[2] == b'/' || bytes[2] == b'\\')
    {
        return Some(ManifestViolation::PathEscape);
    }
    if path.contains('\n') || path.contains('\0') {
        return Some(ManifestViolation::UnsafeName);
    }
    let normalized;
    let segments: &str = if win32 {
        normalized = path.replace('\\', "/");
        &normalized
    } else {
        path
    };
    let escapes = segments
        .split('/')
        .any(|segment| segment.is_empty() || segment == "." || segment == "..");
    escapes.then_some(ManifestViolation::PathEscape)
}

/// Builds a directory hash from `(relative_path, file_digest)` pairs,
/// sorting them by UTF-16 code units the way `comparePaths` does. Each line
/// is `<utf16 length>:<path><64 hex>`, concatenated without separators.
#[derive(Debug, Default)]
pub(crate) struct DirectoryManifest {
    lines: Vec<(String, String)>,
    size_bytes: u64,
}

impl DirectoryManifest {
    /// Adds one leaf's relative path, file digest, and byte length.
    pub(crate) fn push(&mut self, relative_path: &str, file_digest: String, len: u64) {
        self.lines.push((relative_path.to_string(), file_digest));
        self.size_bytes += len;
    }

    /// The final `(content_hash, size_bytes)`.
    #[must_use]
    pub(crate) fn finish(mut self) -> (String, u64) {
        self.lines
            .sort_by(|(left, _), (right, _)| cmp_utf16(left, right));
        let mut manifest = String::new();
        for (path, digest) in &self.lines {
            manifest.push_str(&utf16_len(path).to_string());
            manifest.push(':');
            manifest.push_str(path);
            manifest.push_str(digest);
        }
        (
            sha256_hex(DIRECTORY_HASH_DOMAIN, manifest.as_bytes()),
            self.size_bytes,
        )
    }
}

/// `whitespaceDigest`: SHA-256 of the decoded text with every ECMAScript
/// whitespace character removed.
#[must_use]
pub(crate) fn whitespace_digest(bytes: &[u8]) -> String {
    let text = strip_js_whitespace(&text_decoder_decode(bytes));
    sha256_hex(b"", text.as_bytes())
}

/// `combineWhitespaceDigests`: order-independent, so entries are sorted by
/// `localeCompare` on their paths first. Duplicate entries are hashed as
/// given — the TypeScript hash pass can record one file twice when two leaf
/// names resolve to it, and that is part of the digest.
#[must_use]
pub(crate) fn combine_whitespace_digests(entries: &[(String, String)]) -> String {
    let mut sorted: Vec<&(String, String)> = entries.iter().collect();
    // Stable, like `Array.prototype.sort`, so equal paths keep read order.
    sorted.sort_by(|(left, _), (right, _)| locale_compare(left, right));
    let mut hasher = Sha256::new();
    hasher.update(WHITESPACE_HASH_DOMAIN);
    for (path, digest) in sorted {
        hasher.update(path.as_bytes());
        hasher.update(b"\0");
        hasher.update(digest.as_bytes());
        hasher.update(b"\n");
    }
    hex(&hasher.finalize())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_advertised_directory_domain_version_is_read_from_the_domain_itself() {
        assert_eq!(DIRECTORY_HASH_DOMAIN_VERSION, 2);
        assert_eq!(
            directory_hash_domain_version(b"mangostudio/library/dir/v3\0"),
            3
        );
        assert_eq!(
            directory_hash_domain_version(b"mangostudio/library/dir/v255\0"),
            255
        );
    }

    #[test]
    #[should_panic(expected = "must end in /v<n>")]
    fn a_directory_domain_without_a_version_suffix_is_refused() {
        let _ = directory_hash_domain_version(b"mangostudio/library/dir\0");
    }

    #[test]
    #[should_panic(expected = "1..=255")]
    fn a_directory_domain_version_the_schema_cannot_carry_is_refused() {
        let _ = directory_hash_domain_version(b"mangostudio/library/dir/v256\0");
    }

    #[test]
    fn file_and_directory_domains_never_collide_on_a_forged_manifest() {
        let digest = hash_file_bytes(b"x");
        let mut manifest = DirectoryManifest::default();
        manifest.push("a", digest.clone(), 1);
        let (directory, _) = manifest.finish();
        let forged = format!("1:a{digest}");
        assert_ne!(
            directory,
            hash_file_bytes(forged.as_bytes()),
            "a file spelling a one-entry manifest must not hash like the directory it describes"
        );
    }

    #[test]
    fn relative_path_violations_match_hash_ts() {
        assert_eq!(
            relative_path_violation("a\nb", false),
            Some(ManifestViolation::UnsafeName)
        );
        assert_eq!(
            relative_path_violation("a/../b", false),
            Some(ManifestViolation::PathEscape)
        );
        assert_eq!(relative_path_violation("a\\..\\b", false), None);
        assert_eq!(
            relative_path_violation("a\\..\\b", true),
            Some(ManifestViolation::PathEscape)
        );
        assert_eq!(
            relative_path_violation("C:/x", true),
            Some(ManifestViolation::PathEscape)
        );
        for escaping in ["a/./b", "a//b", ".", "a/.."] {
            assert_eq!(
                relative_path_violation(escaping, false),
                Some(ManifestViolation::PathEscape),
                "expected {escaping:?} to be refused as a path escape"
            );
        }
        // Only `^[A-Za-z]:[\\/]` is a drive prefix (the regex hash.ts uses):
        // names that merely resemble one stay hashable.
        for name in ["ab\\c", "C:x", "a:b", "1:/x"] {
            assert_eq!(
                relative_path_violation(name, true),
                None,
                "expected {name:?} to be a safe win32 name | received a violation"
            );
        }
        assert_eq!(
            relative_path_violation("C:\\x", true),
            Some(ManifestViolation::PathEscape)
        );
        assert_eq!(
            relative_path_violation("", false),
            Some(ManifestViolation::PathEscape)
        );
    }
}
